designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+  Expriation: 3
+  empty: 2
+}
+```
+
+Assuming zero unhealthy nodes:
+**First Equation**
+- drifted = 10 - 3: Disruption allowed as drifted bucket isn't full.
+- underutilized = 15 - 6: Disruption allowed since 15 - 6 > 0.
+- Default = 5 - expired - empty = 0: No disruption allowed for reasons other than consolidation and drift.
+
+**Second Equation**
+- Total Disrupting = 3 + 6 + 5 = 14.
+- drifted = 10 - 14: No disruption allowed due to drift.
+- underutilized = 15 - 14: Disruption allowed for at least one node due to consolidation.
+- Default: 5 - 14: No disruption allowed for other methods.
+
+#### Decision
+The second equation simplifies reasoning for cluster operators. A node can be tagged for multiple disruption reasons (e.g., empty, drifted, underutilized), making it challenging to precisely calculate in-flight disruptions. The second equation streamlines this process by providing a clearer view of the overall disruption impact.
+### Q: How should we handle an unspecified default reason? 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drift, underutilized]
+    schedule: "* * * * *"
+  - nodes: 5 
+    reasons: [empty] 
+    schedule: "* * * * *" 
+```
+In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  
+The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).
+Lets choose total number of nodes, since this allows the user to also specify periods where no nodes are to be disrupted of a particular type of disruption, and makes more sense with the existing karpenter behavior today.
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget - Recommended
+This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
+#### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of methods for disruption that apply to this budget. If Reasons is not set, this budget applies to all methods.
+      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
+      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
+      // allowed reasons are "consolidation", "expired", "empty", "drift"
```

Comment:
```suggestion
      // allowed reasons are "underutilized", "expired", "empty", "drift"
```

Consolidation _has_ a policy that covers empty and underutilized so I think that we should rename this from underutilized to empty. Realistically, this is the difference between the number of nodes that we can disrupt that aren't fully empty vs. ones that are, which I think is what users really care about here.

Reply:
The other question that I have here: do we need something that describes the difference between replace and delete? Maybe this is an extension into the future, but I could imagine _this_ is what users care about here too. e.g. When you disrupt my node, is it going to go onto existing capacity (e.g. I can trust that it will run because pods are already there) or is it going to go onto new capacity? 

The reason that I call this out is that when we are doing a consolidation replace for a NodePool that has drifted configuration, the replacement node is still going to have the new configuration, even if we have fully blocked drift, which may be what the user was trying to avoid in the first place. Have we thought about this case and if this is what users want/expect?